pkg/dataplane/reloadCredentials.go

```diff
@@ -0,0 +1,154 @@
+package dataplane
+
+import (
+	"context"
+	"errors"
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/msi-dataplane/pkg/dataplane/swagger"
+	"github.com/fsnotify/fsnotify"
+)
+
+var (
+	// Errors returned when reloading credentials
+	errCreateFileWatcher = errors.New("failed to create file watcher")
+	errAddFileToWatcher  = errors.New("failed to add credentialFile to file watcher")
+	errLoadCredentials   = errors.New("failed to load credentials from file")
+)
+
+type reloadingCredential struct {
+	currentValue *azidentity.ClientCertificateCredential
+	cloud        string
+	lock         *sync.RWMutex
+	logger       *log.Logger
+}
+
+type Option func(*reloadingCredential)
+
+// WithLogger sets a custom logger for the reloadingCredential.
+// This can be useful for debugging or logging purposes.
+func WithLogger(logger *log.Logger) Option {
+	return func(c *reloadingCredential) {
+		c.logger = logger
+	}
+}
+
+// NewUserAssignedIdentityCredential creates a new reloadingCredential for a user-assigned identity.
+// ctx is used to manage the lifecycle of the credential, allowing for cancellation and timeouts.
+// cloud specifies the cloud environment.
+// credentialPath is the path to the credential file.
+// opts allows for additional configuration, such as setting a custom logger.
+//
+// The function ensures that a valid token is loaded before returning the credential.
+// It also starts a background process to watch for changes to the credential file and reloads it as necessary.
+// In any case the maximum time that we wait before reload is six hours.
```

Comment:
Sorry, I think my other comment got lost in the thread - let's expose the duration here as an `Option` as well. Defaulting to your six hour period sounds good.